### PR TITLE
Harden watcher for more perf, use exec and watch from worker

### PR DESCRIPTION
**Pull request by theopolis**
* The watcher/worker model on OSX was flawed in that it forked without execing itself. 
* The worker now execs after setting an environment variable indicating the daemon should skip startup checks.
* `OSQUERYD_WORKER` indicates a worker should skip setting pid, checking DBInstance, and some logging.
* The performance monitoring of the worker was latent in most cases because of the implicit requirement to start an in-memory SQL database.
* This now skips the SQL and emits data directly from the processes table plugin using `Registry::call`. 
* On OSX CPU time is measured with different precision that Linux, the watcher now monitors utilization accordingly.
* The precision on OSX would cause an integer overflow in abort the watch (thus worker) within a reasonable amount of time.
* The worker now starts a "watcher watcher" process
* If the worker's parent changes (indicating it has become abandoned) it will crash itself.
* Processes table now select on pid more effectively.
* Some added functions to QueryContext.
* The DBInstance class has a new `checkDB` method that will release the instance after checking sanity.
*

### CODE REVIEW

**Comment by theopolis** [comment]
Implements: https://github.com/facebook/osquery/issues/703

File: osquery/tables/system/linux/processes.cpp [hunk 1]
```diff
@@ -140,8 +140,12 @@ QueryData genProcesses(QueryContext& context) {
 
   // Populate proc struc for each process.
   while ((proc_info = readproc(proc, NULL))) {
-    Row r;
+    if (!context.constraints["pid"].matches<int>(proc_info->tid)) {
+      // Optimize by not searching when a pid is a constraint.
+      continue;
```
Comment (theopolis) [inline osquery/tables/system/linux/processes.cpp]: Would cause a nasty memory leak.

**Comment by theopolis** [comment]
Implements #692 too

File: include/osquery/database/db_handle.h [hunk 2]
```diff
@@ -88,6 +88,16 @@ class DBHandle {
   static std::shared_ptr<DBHandle> getInstance();
 
   /**
+   * @brief Check the sanify of the database configuration options
```
Comment (marpaia) [inline include/osquery/database/db_handle.h]: sanity*

File: include/osquery/database/db_handle.h [hunk 3]
```diff
@@ -88,6 +88,16 @@ class DBHandle {
...
+   *
+   * Create a handle to the backing store using the database configuration.
+   * Catch any instance creation exceptions and release the handle immediately.
+   *
+   * @return Success if a handle was created without error.
+   */
+  static bool checkDB();
```
Comment (marpaia) [inline include/osquery/database/db_handle.h]: obligatory comment about using booleans to represent status'

File: osquery/core/watcher.cpp [hunk 4]
```diff
@@ -14,22 +14,46 @@
...
+    // Seconds of tolerable sustained latency.
+    {LATENCY_LIMIT, {5, 5, 3}},
+    // How often to poll for performance limit violations.
+    {INTERVAL, {3, 3, 3}}, };
+
+DEFINE_osquery_flag(
+    int32,
+    watchdog_level,
+    1,
+    "Performance limit level (0=loose, 1=normal, 2=restrictive)");
```
Comment (marpaia) [inline osquery/core/watcher.cpp]: nice!

**Comment by marpaia** [comment]
image

**Comment by javuto** [comment]
@osqueryer, ok to test

**Comment by javuto** [comment]
@osqueryer, ok to test

**Comment by javuto** [comment]
@osqueryer, ok to test

**Comment by theopolis** [comment]
ok to test

**Pull request merged**